File: java/Ice/async/WorkQueue.java
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// **********************************************************************

import Demo.*;

public class WorkQueue extends Thread
{
    class FutureEntry
    {
        java.util.concurrent.CompletableFuture<Void> future;
        int delay;
    }

    @Override
    public synchronized void run()
    {
        while(!_done)
        {
            if(_futures.size() == 0)
            {
                try
                {
                    wait();
                }
                catch(java.lang.InterruptedException ex)
                {
                }
            }

            if(_futures.size() != 0)
            {
                //
                // Get next work item.
                //
                FutureEntry entry = _futures.getFirst();

                //
                // Wait for the amount of time indicated in delay to
                // emulate a process that takes a significant period of
                // time to complete.
                //
                try
                {
                    wait(entry.delay);
                }
                catch(java.lang.InterruptedException ex)
                {
                }

                if(!_done)
                {
                    //
                    // Print greeting and send response.
                    //
                    _futures.removeFirst();
                    System.err.println("Belated Hello World!");
                    entry.future.complete((Void)null);
                }
            }
        }

        //
        // Throw exception for any outstanding requests.
        //
        for(FutureEntry p : _futures)
        {
            p.future.completeExceptionally(new RequestCanceledException());
        }
    }

    public synchronized void add(java.util.concurrent.CompletableFuture<Void> future, int delay)
    {
        if(!_done)
        {
            //
            // Add the work item.
            //
            FutureEntry entry = new FutureEntry();
            entry.future = future;
            entry.delay = delay;

            if(_futures.size() == 0)
            {
                notify();
            }
            _futures.add(entry);
        }
        else
        {
            //
            // Destroyed, throw exception.
            //
            future.completeExceptionally(new RequestCanceledException());
        }
    }

    public synchronized void _destroy()                  // Thread.destroy is deprecated.
    {
        _done = true;
        notify();
    }

    private java.util.LinkedList<FutureEntry> _futures = new java.util.LinkedList<>();
    private boolean _done = false;
}

File: cpp/Ice/async/WorkQueue.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#include "WorkQueue.h"
#include <Ice/Ice.h>
#include <iostream>

using namespace std;

void
WorkQueue::start()
{
    thread t([this]() { this->run(); });
    _thread = std::move(t);
}

void
WorkQueue::join()
{
    if (_thread.joinable())
    {
        _thread.join();
    }
}

void
WorkQueue::run()
{
    unique_lock<mutex> lock(_mutex);

    while (!_done)
    {
        if (_callbacks.empty())
        {
            _condition.wait(lock);
        }

        if (!_done && !_callbacks.empty())
        {
            //
            // Get next work item.
            //
            CallbackEntry entry = _callbacks.front();

            //
            // Wait for the amount of time indicated in delay to
            // emulate a process that takes a significant period of
            // time to complete.
            //
            _condition.wait_for(lock, chrono::milliseconds(get<0>(entry)));

            if (!_done)
            {
                //
                // Print greeting and send response.
                //
                _callbacks.pop_front();
                cout << "Belated Hello World!" << endl;
                auto& response = get<1>(entry);
                response();
            }
        }
    }

    //
    // Throw exception for any outstanding requests.
    //
    for (auto& entry : _callbacks)
    {
        get<2>(entry)(make_exception_ptr(Demo::RequestCanceledException()));
    }
}

void
WorkQueue::add(int delay, function<void()> response, function<void(exception_ptr)> error)
{
    const unique_lock<mutex> lock(_mutex);

    if (!_done)
    {
        //
        // Add work item.
        //
        if (_callbacks.size() == 0)
        {
            _condition.notify_one();
        }
        _callbacks.emplace_back(delay, std::move(response), std::move(error));
    }
    else
    {
        //
        // Destroyed, throw exception.
        //
        error(make_exception_ptr(Demo::RequestCanceledException()));
    }
}

void
WorkQueue::destroy()
{
    const unique_lock<mutex> lock(_mutex);

    //
    // Set done flag and notify.
    //
    _done = true;
    _condition.notify_one();
}
